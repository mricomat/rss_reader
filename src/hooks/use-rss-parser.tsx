import { useEffect, useState } from 'react';
import { DOMParser } from 'xmldom';

import useCache from './use-cache';
import { useFetch } from './use-fetch';

import { IRss, IRssLink, IRssEnclosure, IRssItem } from 'src/types/data';
import RssUtils from 'src/utils/rss';
import { buildRequestId } from 'src/utils/text';

const defRssValues: IRss = {
  title: '',
  links: [],
  description: '',
  items: [],
};

export const useRssParser = (
  url: string = '',
  text: string = '',
  persist: boolean = false,
  cacheLife: number = 24 * 60000
): IRss => {
  const { doFetch, data } = useFetch<any>(persist, cacheLife);
  const cache = useCache<IRss>({ persist, cacheLife });

  const [rssData, setRssData] = useState<IRss>(defRssValues);

  useEffect(() => {
    if (url !== '') doFetchText();
    else parseText(text);
  }, [url, text]);

  useEffect(() => {
    if (data && typeof data === 'string') {
      parseText(data);
    }
  }, [data]);

  const doFetchText = async () => {
    const requestId = buildRequestId({ url });

    if (await cache.has(requestId)) {
      const oldRssData = await cache.get(requestId);
      if (oldRssData) setRssData(oldRssData);
    } else {
      doFetch('get', url, {});
    }
  };

  const parseText = async (newText: string) =>
    new Promise<void>((resolve, reject) => {
      const document = new DOMParser({
        errorHandler: (_level, msg) => {
          reject(msg);
        },
      }).parseFromString(newText, 'text/xml');

      const newRssData: IRss = { ...mapChannelFields(document), items: [...mapItems(document)] };
      setCache(buildRequestId({ url }), newRssData);

      setRssData(newRssData);
      resolve();
    });

  const setCache = async (requestId: string, newRssData: IRss) => {
    await cache.set(requestId, newRssData);
  };

  const getChannelTitle = (node): string => RssUtils.getElementTextContent(node, 'title');

  const getChannelLinks = (node): IRssLink[] => {
    const links = RssUtils.getChildElements(node, 'link');

    return links.map(link => ({
      url: link.textContent,
      rel: link.getAttribute('rel'),
    }));
  };

  const getChannelDescription = (node): string => RssUtils.getElementTextContent(node, 'description');

  const getItemTitle = (node): string => RssUtils.getElementTextContent(node, 'title');

  const getItemLinks = (node): IRssLink[] => {
    const links = RssUtils.getChildElements(node, 'link');

    return links.map(link => ({
      url: link.textContent,
      rel: link.getAttribute('rel'),
    }));
  };

  const getItemDescription = (node): string => RssUtils.getElementTextContent(node, 'description');

  const getItemId = (node): string => RssUtils.getElementTextContent(node, 'guid');

  const getItemEnclosures = (node): IRssEnclosure[] => {
    const enclosures = RssUtils.getChildElements(node, 'enclosure');

    return enclosures.map(enclosure => ({
      url: enclosure.getAttribute('url'),
      length: enclosure.getAttribute('length'),
      mimeType: enclosure.getAttribute('type'),
    }));
  };

  const mapChannelFields = document => {
    if (document) {
      const channelNodes = RssUtils.getElements(document, 'channel');

      if (!channelNodes || channelNodes.length === 0) {
        throw new Error('Could not find channel node');
      }

      const channelNode = channelNodes[0];

      return {
        title: getChannelTitle(channelNode),
        links: getChannelLinks(channelNode),
        description: getChannelDescription(channelNode),
      };
    }
    return {};
  };

  const mapItems = (document): IRssItem[] => {
    if (document) {
      const itemNodes = RssUtils.getElements(document, 'item');

      return itemNodes.map(item => ({
        title: getItemTitle(item),
        links: getItemLinks(item),
        description: getItemDescription(item),
        id: getItemId(item),
        enclosures: getItemEnclosures(item),
      }));
    }
    return [];
  };

  return {
    ...rssData,
  };
};
