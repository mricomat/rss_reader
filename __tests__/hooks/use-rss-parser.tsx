import { renderHook } from '@testing-library/react-hooks';
import { waitFor } from '@testing-library/react-native';

import { useRssParser } from 'src/hooks/use-rss-parser';
import validText from '../../__mocks__/rss-samples/valid-document';

describe('when hook/use-rss-cache', () => {
  it('given text should return correctly the data', async () => {
    const { result } = renderHook(() => useRssParser('', validText, false, 24 * 600000));
    const { title, description, items } = result.current;

    await waitFor(
      () => {
        expect(title).toEqual('NASA Breaking News');
        expect(description).toEqual('A RSS news feed containing the latest NASA news articles and press releases.');
        expect(items[0].title).toEqual('NASA Extends Cyclone Global Navigation Satellite System Mission');
        expect(items[0].description).toEqual(
          'NASA has awarded a contract to the University of Michigan for the Cyclone Global Navigation Satellite System (CYGNSS) for mission operations and closeout.'
        );
        expect(items[0].enclosures[0].url).toEqual(
          'http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/cygnss20191007-home.jpg?itok=jcr5f5XK'
        );
        expect(items[0].links[0].url).toEqual(
          'http://www.nasa.gov/press-release/nasa-extends-cyclone-global-navigation-satellite-system-mission'
        );
        expect(result).toMatchSnapshot();
      },
      { timeout: 1000 }
    );
  });
});
