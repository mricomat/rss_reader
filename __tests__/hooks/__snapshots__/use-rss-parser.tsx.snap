// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when hook/use-rss-cache given text should return correctly the data 1`] = `
Object {
  "all": Array [
    Object {
      "description": "",
      "items": Array [],
      "links": Array [],
      "title": "",
    },
    Object {
      "description": "A RSS news feed containing the latest NASA news articles and press releases.",
      "items": Array [
        Object {
          "description": "NASA has awarded a contract to the University of Michigan for the Cyclone Global Navigation Satellite System (CYGNSS) for mission operations and closeout.",
          "enclosures": Array [
            Object {
              "length": "428731",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/cygnss20191007-home.jpg?itok=jcr5f5XK",
            },
          ],
          "id": "http://www.nasa.gov/press-release/nasa-extends-cyclone-global-navigation-satellite-system-mission",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/nasa-extends-cyclone-global-navigation-satellite-system-mission",
            },
          ],
          "title": "NASA Extends Cyclone Global Navigation Satellite System Mission",
        },
        Object {
          "description": "Northrop Grumman’s uncrewed Cygnus spacecraft is scheduled to depart the International Space Station Tuesday, June 29, more than four months after delivering nearly 8,000 pounds of supplies, scientific investigations, commercial products, hardware, and other cargo to the orbital laboratory.",
          "enclosures": Array [
            Object {
              "length": "1822735",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/iss064e036104_orig.jpg?itok=4dIZWLrd",
            },
          ],
          "id": "http://www.nasa.gov/press-release/nasa-to-air-northrop-grumman-cygnus-departure-from-space-station",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/nasa-to-air-northrop-grumman-cygnus-departure-from-space-station",
            },
          ],
          "title": "NASA to Air Northrop Grumman Cygnus Departure from Space Station",
        },
        Object {
          "description": "Two astronauts will venture outside the International Space Station Friday, June 25, for a third spacewalk to continue power system upgrades that are already increasing output and proving the technology that will enable NASA’s future Gateway lunar outpost.",
          "enclosures": Array [
            Object {
              "length": "1652402",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/iss065e121336.jpeg?itok=j5cb3D8h",
            },
          ],
          "id": "http://www.nasa.gov/press-release/nasa-to-air-third-spacewalk-to-install-new-station-solar-arrays",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/nasa-to-air-third-spacewalk-to-install-new-station-solar-arrays",
            },
          ],
          "title": "NASA to Air Third Spacewalk to Install New Station Solar Arrays",
        },
        Object {
          "description": "NASA has awarded Aerodyne-SGT Engineering Services LLC (ASES) of Cape Canaveral, Florida, for a contract to provide engineering services for spaceflight and ground systems.",
          "enclosures": Array [
            Object {
              "length": "189751",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/nasa-logo-web-rgb_0.jpg?itok=mrBnB_c9",
            },
          ],
          "id": "http://www.nasa.gov/press-release/nasa-selects-engineering-services-contractor",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/nasa-selects-engineering-services-contractor",
            },
          ],
          "title": "NASA Selects Engineering Services Contractor",
        },
        Object {
          "description": "NASA Deputy Administrator Pam Melroy took office Monday after she was given the oath of office by NASA Administrator Bill Nelson during a ceremony at the Mary W. Jackson NASA Headquarters building in Washington.",
          "enclosures": Array [
            Object {
              "length": "583330",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/51261357997_5889560e6a_k.jpeg?itok=fIG440Nc",
            },
          ],
          "id": "http://www.nasa.gov/press-release/pam-melroy-sworn-in-as-nasa-deputy-administrator",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/pam-melroy-sworn-in-as-nasa-deputy-administrator",
            },
          ],
          "title": "Pam Melroy Sworn in as NASA Deputy Administrator",
        },
        Object {
          "description": "The following are statements from Pam Melroy and Administrator Bill Nelson on Thursday’s U.S. Senate confirmation of Melroy as deputy administrator of the National Aeronautics and Space Administration",
          "enclosures": Array [
            Object {
              "length": "1126368",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/9730024457_617bec9c42_o.jpg?itok=47gfEOQA",
            },
          ],
          "id": "http://www.nasa.gov/press-release/statements-on-pam-melroy-s-senate-confirmation-as-nasa-deputy-administrator",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/statements-on-pam-melroy-s-senate-confirmation-as-nasa-deputy-administrator",
            },
          ],
          "title": "Statements on Pam Melroy’s Senate Confirmation as NASA Deputy Administrator",
        },
        Object {
          "description": "Cleaner-burning jet fuels made from sustainable sources can produce 50%-70% fewer ice crystal contrails at cruising altitude, reducing aviation’s impact on the environment, according to research conducted by NASA and the German Aerospace Center (DLR).",
          "enclosures": Array [
            Object {
              "length": "4301761",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/contrails-photo1.jpg?itok=S0Ga2B4p",
            },
          ],
          "id": "http://www.nasa.gov/press-release/nasa-dlr-study-finds-sustainable-aviation-fuel-can-reduce-contrails",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/nasa-dlr-study-finds-sustainable-aviation-fuel-can-reduce-contrails",
            },
          ],
          "title": "NASA-DLR Study Finds Sustainable Aviation Fuel Can Reduce Contrails",
        },
        Object {
          "description": "NASA Administrator Bill Nelson released the following statement Thursday regarding the China National Space Agency’s launch of the first crew to its Tiangong space station:",
          "enclosures": Array [
            Object {
              "length": "189751",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/nasa-logo-web-rgb_0.jpg?itok=mrBnB_c9",
            },
          ],
          "id": "http://www.nasa.gov/press-release/nasa-administrator-statement-on-china-crewed-launch",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/nasa-administrator-statement-on-china-crewed-launch",
            },
          ],
          "title": "NASA Administrator Statement on China Crewed Launch",
        },
        Object {
          "description": "La NASA lanza la Misión Equity (equidad, en español), un esfuerzo integral para evaluar la expansión y modificación de programas, compras de suministros, becas, y políticas de la agencia  y examinar qué obstáculos y desafíos potenciales existen para las comunidades que históricamente están subrepresentadas y desatendidas.",
          "enclosures": Array [
            Object {
              "length": "1103939",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/la_mision_equity3.jpg?itok=Fru9J0X7",
            },
          ],
          "id": "http://www.nasa.gov/press-release/la-nasa-lanza-la-misi-n-equity-y-solicita-comentarios-del-p-blico-para-ampliar-el",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/la-nasa-lanza-la-misi-n-equity-y-solicita-comentarios-del-p-blico-para-ampliar-el",
            },
          ],
          "title": "La NASA lanza la misión Equity y solicita comentarios del público para ampliar el acceso",
        },
        Object {
          "description": "NASA is launching Mission Equity, a comprehensive effort to assess expanding and modifying agency programs, procurements, grants, and policies, and examine what potential barriers and challenges exist for communities that are historically underrepresented and underserved.",
          "enclosures": Array [
            Object {
              "length": "189751",
              "mimeType": "image/jpeg",
              "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/nasa-logo-web-rgb.jpg?itok=36eU9WU7",
            },
          ],
          "id": "http://www.nasa.gov/press-release/nasa-launches-mission-equity-seeks-public-input-to-broaden-access",
          "links": Array [
            Object {
              "rel": "",
              "url": "http://www.nasa.gov/press-release/nasa-launches-mission-equity-seeks-public-input-to-broaden-access",
            },
          ],
          "title": "NASA Launches Mission Equity, Seeks Public Input to Broaden Access",
        },
      ],
      "links": Array [
        Object {
          "rel": "",
          "url": "http://www.nasa.gov/",
        },
      ],
      "title": "NASA Breaking News",
    },
  ],
  "current": Object {
    "description": "A RSS news feed containing the latest NASA news articles and press releases.",
    "items": Array [
      Object {
        "description": "NASA has awarded a contract to the University of Michigan for the Cyclone Global Navigation Satellite System (CYGNSS) for mission operations and closeout.",
        "enclosures": Array [
          Object {
            "length": "428731",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/cygnss20191007-home.jpg?itok=jcr5f5XK",
          },
        ],
        "id": "http://www.nasa.gov/press-release/nasa-extends-cyclone-global-navigation-satellite-system-mission",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/nasa-extends-cyclone-global-navigation-satellite-system-mission",
          },
        ],
        "title": "NASA Extends Cyclone Global Navigation Satellite System Mission",
      },
      Object {
        "description": "Northrop Grumman’s uncrewed Cygnus spacecraft is scheduled to depart the International Space Station Tuesday, June 29, more than four months after delivering nearly 8,000 pounds of supplies, scientific investigations, commercial products, hardware, and other cargo to the orbital laboratory.",
        "enclosures": Array [
          Object {
            "length": "1822735",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/iss064e036104_orig.jpg?itok=4dIZWLrd",
          },
        ],
        "id": "http://www.nasa.gov/press-release/nasa-to-air-northrop-grumman-cygnus-departure-from-space-station",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/nasa-to-air-northrop-grumman-cygnus-departure-from-space-station",
          },
        ],
        "title": "NASA to Air Northrop Grumman Cygnus Departure from Space Station",
      },
      Object {
        "description": "Two astronauts will venture outside the International Space Station Friday, June 25, for a third spacewalk to continue power system upgrades that are already increasing output and proving the technology that will enable NASA’s future Gateway lunar outpost.",
        "enclosures": Array [
          Object {
            "length": "1652402",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/iss065e121336.jpeg?itok=j5cb3D8h",
          },
        ],
        "id": "http://www.nasa.gov/press-release/nasa-to-air-third-spacewalk-to-install-new-station-solar-arrays",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/nasa-to-air-third-spacewalk-to-install-new-station-solar-arrays",
          },
        ],
        "title": "NASA to Air Third Spacewalk to Install New Station Solar Arrays",
      },
      Object {
        "description": "NASA has awarded Aerodyne-SGT Engineering Services LLC (ASES) of Cape Canaveral, Florida, for a contract to provide engineering services for spaceflight and ground systems.",
        "enclosures": Array [
          Object {
            "length": "189751",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/nasa-logo-web-rgb_0.jpg?itok=mrBnB_c9",
          },
        ],
        "id": "http://www.nasa.gov/press-release/nasa-selects-engineering-services-contractor",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/nasa-selects-engineering-services-contractor",
          },
        ],
        "title": "NASA Selects Engineering Services Contractor",
      },
      Object {
        "description": "NASA Deputy Administrator Pam Melroy took office Monday after she was given the oath of office by NASA Administrator Bill Nelson during a ceremony at the Mary W. Jackson NASA Headquarters building in Washington.",
        "enclosures": Array [
          Object {
            "length": "583330",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/51261357997_5889560e6a_k.jpeg?itok=fIG440Nc",
          },
        ],
        "id": "http://www.nasa.gov/press-release/pam-melroy-sworn-in-as-nasa-deputy-administrator",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/pam-melroy-sworn-in-as-nasa-deputy-administrator",
          },
        ],
        "title": "Pam Melroy Sworn in as NASA Deputy Administrator",
      },
      Object {
        "description": "The following are statements from Pam Melroy and Administrator Bill Nelson on Thursday’s U.S. Senate confirmation of Melroy as deputy administrator of the National Aeronautics and Space Administration",
        "enclosures": Array [
          Object {
            "length": "1126368",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/9730024457_617bec9c42_o.jpg?itok=47gfEOQA",
          },
        ],
        "id": "http://www.nasa.gov/press-release/statements-on-pam-melroy-s-senate-confirmation-as-nasa-deputy-administrator",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/statements-on-pam-melroy-s-senate-confirmation-as-nasa-deputy-administrator",
          },
        ],
        "title": "Statements on Pam Melroy’s Senate Confirmation as NASA Deputy Administrator",
      },
      Object {
        "description": "Cleaner-burning jet fuels made from sustainable sources can produce 50%-70% fewer ice crystal contrails at cruising altitude, reducing aviation’s impact on the environment, according to research conducted by NASA and the German Aerospace Center (DLR).",
        "enclosures": Array [
          Object {
            "length": "4301761",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/contrails-photo1.jpg?itok=S0Ga2B4p",
          },
        ],
        "id": "http://www.nasa.gov/press-release/nasa-dlr-study-finds-sustainable-aviation-fuel-can-reduce-contrails",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/nasa-dlr-study-finds-sustainable-aviation-fuel-can-reduce-contrails",
          },
        ],
        "title": "NASA-DLR Study Finds Sustainable Aviation Fuel Can Reduce Contrails",
      },
      Object {
        "description": "NASA Administrator Bill Nelson released the following statement Thursday regarding the China National Space Agency’s launch of the first crew to its Tiangong space station:",
        "enclosures": Array [
          Object {
            "length": "189751",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/nasa-logo-web-rgb_0.jpg?itok=mrBnB_c9",
          },
        ],
        "id": "http://www.nasa.gov/press-release/nasa-administrator-statement-on-china-crewed-launch",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/nasa-administrator-statement-on-china-crewed-launch",
          },
        ],
        "title": "NASA Administrator Statement on China Crewed Launch",
      },
      Object {
        "description": "La NASA lanza la Misión Equity (equidad, en español), un esfuerzo integral para evaluar la expansión y modificación de programas, compras de suministros, becas, y políticas de la agencia  y examinar qué obstáculos y desafíos potenciales existen para las comunidades que históricamente están subrepresentadas y desatendidas.",
        "enclosures": Array [
          Object {
            "length": "1103939",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/la_mision_equity3.jpg?itok=Fru9J0X7",
          },
        ],
        "id": "http://www.nasa.gov/press-release/la-nasa-lanza-la-misi-n-equity-y-solicita-comentarios-del-p-blico-para-ampliar-el",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/la-nasa-lanza-la-misi-n-equity-y-solicita-comentarios-del-p-blico-para-ampliar-el",
          },
        ],
        "title": "La NASA lanza la misión Equity y solicita comentarios del público para ampliar el acceso",
      },
      Object {
        "description": "NASA is launching Mission Equity, a comprehensive effort to assess expanding and modifying agency programs, procurements, grants, and policies, and examine what potential barriers and challenges exist for communities that are historically underrepresented and underserved.",
        "enclosures": Array [
          Object {
            "length": "189751",
            "mimeType": "image/jpeg",
            "url": "http://www.nasa.gov/sites/default/files/styles/1x1_cardfeed/public/thumbnails/image/nasa-logo-web-rgb.jpg?itok=36eU9WU7",
          },
        ],
        "id": "http://www.nasa.gov/press-release/nasa-launches-mission-equity-seeks-public-input-to-broaden-access",
        "links": Array [
          Object {
            "rel": "",
            "url": "http://www.nasa.gov/press-release/nasa-launches-mission-equity-seeks-public-input-to-broaden-access",
          },
        ],
        "title": "NASA Launches Mission Equity, Seeks Public Input to Broaden Access",
      },
    ],
    "links": Array [
      Object {
        "rel": "",
        "url": "http://www.nasa.gov/",
      },
    ],
    "title": "NASA Breaking News",
  },
  "error": undefined,
}
`;
